pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -454,10 +449,10 @@ func (nc *Controller) doNoScheduleTaintingPass(node *v1.Node) error {
 	// Map node's condition to Taints.
 	var taints []v1.Taint
 	for _, condition := range node.Status.Conditions {
-		if taint, found := nodeConditionToTaintKeyStatusMap[condition.Type]; found {
-			if condition.Status == taint.noScheduleStatus {
+		if taintMap, found := nodeConditionToTaintKeyStatusMap[condition.Type]; found {
+			if taintKey, found := taintMap[condition.Status]; found {
```

Comment:
If you look at line 70, it seems that NodeUnreachable should be a NoExecute taint, i.e. causes eviction of running pods. If that's true, we should not always use NoSchedule here, at line 455.

Reply:
`NoSchedule` is used for scheduling, which is different. This fix can address issue that scheduler still schedule pods to NotReady node.
Another question, did you try shutting down node and see whether both `node.kubernetes.io/not-ready:NoSchedule` and `node.kubernetes.io/not-ready:NoExecute` taint are there?  Code https://github.com/kubernetes/kubernetes/blob/master/pkg/util/taints/taints.go#L277 seems did not do the job.

Reply:
@bsalamat `UnreachableTaintTemplate` in line 70 (a.k.a unreachable/NoExecute taint) is applied when `TaintBasedEvictions` is enabled - although the name is ambiguous.

I think the idea here is to differentiate NoSchedule and NoExecute taint:
- TaintBasedEvictions controls `not-ready/NoExecute` and `unreachable/NoExecute`
- TaintNodesByCondition controls `xyz/NoSchedule`

And I found another bug that when TaintBasedEvictions and TaintNodesByCondition are both enabled, `xyz/NoSchedule` taint will **overwrite** existing `xyz/NoExecute` taints. It's because the logic below only check if the taintKey is found or not, regardless the effect:
https://github.com/kubernetes/kubernetes/blob/7a9a30a4f7eae900bffd4b477a431a999de350dd/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L474-L483

In latest commit, it's resolved.

So right now when TaintBasedEvictions and TaintNodesByCondition are both enabled, the behavior is like this:

- create a 2-node cluster
- create a deployment with 8 replicas - 4 landed on node1, 4 landed on node2
- break node1's network connectivity to apiserver
- node1 is applied with following taints:
    ```yaml
    spec:
      taints:
      - effect: NoSchedule
        key: node.kubernetes.io/unreachable
        timeAdded: 2018-08-23T19:21:58Z
      - effect: NoExecute
        key: node.kubernetes.io/unreachable
        timeAdded: 2018-08-23T19:22:03Z
    ```
- also pods landed on node1 are applied `NoExecute` toleration - means it will be evicted in 300 seconds:
    ```yaml
      tolerations:
      - effect: NoExecute
        key: node.kubernetes.io/not-ready
        operator: Exists
        tolerationSeconds: 300
      - effect: NoExecute
        key: node.kubernetes.io/unreachable
        operator: Exists
        tolerationSeconds: 300
    ```
- 300 seconds later, node lifecycle manager starts to evicts pods are spawn new pods - due to introduced `unreachable/NoSchedule` taint, new pods all landed to node2
    ```yaml
    # k get po -o wide
    NAME                     READY     STATUS        RESTARTS   AGE       IP            NODE          NOMINATED NODE
    pause-54dfd6bf9c-2nwrm   1/1       Running       0          8m9s      10.244.3.4    kube-node-2   <none>
    pause-54dfd6bf9c-6v62c   1/1       Running       0          44s       10.244.3.9    kube-node-2   <none>
    pause-54dfd6bf9c-8c7rh   1/1       Running       0          44s       10.244.3.8    kube-node-2   <none>
    pause-54dfd6bf9c-8nw4s   1/1       Terminating   0          8m9s      10.244.2.4    kube-node-1   <none>
    pause-54dfd6bf9c-b2sng   1/1       Terminating   0          8m9s      10.244.2.6    kube-node-1   <none>
    pause-54dfd6bf9c-bczdb   1/1       Running       0          8m9s      10.244.3.5    kube-node-2   <none>
    pause-54dfd6bf9c-fms72   1/1       Running       0          44s       10.244.3.7    kube-node-2   <none>
    pause-54dfd6bf9c-g7nnv   1/1       Running       0          8m9s      10.244.3.3    kube-node-2   <none>
    pause-54dfd6bf9c-gv8d5   1/1       Running       0          44s       10.244.3.10   kube-node-2   <none>
    pause-54dfd6bf9c-n5njz   1/1       Running       0          8m9s      10.244.3.6    kube-node-2   <none>
    pause-54dfd6bf9c-sfj6h   1/1       Terminating   0          8m9s      10.244.2.3    kube-node-1   <none>
    pause-54dfd6bf9c-zqtdc   1/1       Terminating   0          8m9s      10.244.2.5    kube-node-1   <none>
    # k get deploy
    NAME      DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
    pause     8         8         8            8           8m21s
    # k get rs
    NAME               DESIRED   CURRENT   READY     AGE
    pause-54dfd6bf9c   8         8         8         8m28s
    ```
- the 4 Terminating pods still shows is b/c kubelet in node1 can't talk to apiserver, hence still have those orphan containers living - but they doesn't count into replicas of the deployment. They are cleaned up gracefully after node1 gets recovered.

Reply:
@liqlin2015 yeap, I see both taints. See above comment for more details.

Reply:
> I think the idea here is to differentiate NoSchedule and NoExecute taint:

Regarding `TaintBasedEviction`, node lifecycle already has feature to evict/kill pod when there's some issue to the node before toleration/taint; and then we leverage NoExecute to do that.

Regarding `TaintNodeByCondition`, there's a requirements to make pod tolerating some condition (e.g. network unavailable for who's using host network) which is hard code in scheduler before.


> taintutils.TaintSetFilter

+1

Reply:
Thanks, @Huang-Wei for the detailed explanation.